README.md
# mongoDB datasource for CakePHP

version: 2.0a

This is an experimental branch by [@dizyart](http://github.com/dizyart/) with a major rework in MongodbSource internals. Most important additions are:

* Tons of bugfixes
* Associations support
* Compatibility (PHP 5.x, CakePHP 2.x, Mongo 1.x)
* MongodbSource strictly inherits from DboSource
* Extendable Models (experimental)
* Much mode ([view changelog](./Changelog.md))

Still to come:

* Replicasets testing
* Cross-db Models tests
* Behaviors for proper NoSQL references (only standard SQL references are currently supported)
* Documentation and examples (just look at tests until then)

## Requirements
PHP5, 
pecl mongo (http://php.net/mongo)
CakePHP 2.x

## Installation

this repository should be installed in the same way as any other plugin.

To install the driver for use in a single application:

	cd my/app/Plugin
	git clone git://github.com/ichikaway/cakephp-mongodb.git Mongodb

To install the driver for use in any/multiple application(s)

	# where ROOT is the name of the directory parent to the base index.php of CakePHP.
	cd ROOT/Plugin
	git clone git://github.com/ichikaway/cakephp-mongodb.git Mongodb
	
## Sample Code

To use this DB driver, install (obviously) and define a db source such as follows:

	<?php
	//app/Config/bootstrap.php
	CakePlugin::load('Mongodb');


	// app/Config/database.php
	class DATABASE_CONFIG {
		public $default = array(
			'datasource' => 'Mongodb.MongodbSource',
			'host' => 'localhost',
			'database' => 'blog',
			'port' => 27017,
			'prefix' => '',
			'persistent' => 'true',
			/* optional auth fields
			'login' => 'mongo',	
			'password' => 'awesomeness',
			'replicaset' => array('host' => 'mongodb://hoge:hogehoge@localhost:27021,localhost:27022/blog', 
			                      'options' => array('replicaSet' => 'myRepl')
					     ),
			*/
		);

		// To make sure all tests are passing create the following entry in app/Config/database.php
		public $test = array(
			'datasource' => 'Mongodb.MongodbSource',
			'database' => 'test_mongo',
			'host' => 'localhost',
			'port' => 27017,
		); 
	}

More detail of replicaset in wiki:
https://github.com/ichikaway/cakephp-mongodb/wiki/How-to-connect-to-replicaset-servers


Model files need to have mongoSchema property - or make use of the schemaless behavior. 

Mongo uses a primary key named "\_id" (cannot be renamed). It can be any format you like but if you don't explicitly set it Mongo will use an automatic 24 character (uu)id.

Before you start, you may find it useful to see [a model sample.](http://github.com/ichikaway/mongoDB-Datasource/blob/master/samples/models/post.php)
There are also some sample [controller actions: find,save,delete,deleteAll,updateAll](http://github.com/ichikaway/mongoDB-Datasource/blob/master/samples/controllers/posts_controller.php) note that your controller code needs no specific code to use this datasource.

## Author
Yasushi Ichikawa ([ichikaway](http://twitter.com/ichikaway))

Andy Dawson ([AD7six](http://twitter.com/AD7six))


## Contributors
[Predominant](http://github.com/predominant/) : Cleanup code, add documentation

[Jrbasso](http://github.com/jrbasso/) : Cleanup code

[tkyk](http://github.com/tkyk/) : Fix bug, Add some function.


## Reference
Reference code, Thank you!

[Nate Abele's lithium mongoDB datasource](http://li3.rad-dev.org/)

[Joél Perras' divan](http://github.com/jperras/divan/)



Changelog.md
## MongoDB CakePHP Plugin changelog ##

There has been a lot of changes for v2.0a, since it is a complete rework of the plugin.

Although a big effort was made to retain backward-compatibility, some things have changed and your code _might_ break. This is not a proper migration guideline, but more of a hint.

### Version 2.0a ###

* Added association support (hasMany, belongsTo, hasOne; HABTM only via MongodbAppModel)
* Added Model inheritance (only via MongodbAppModel)
* Default primary key is 'id', the translation to '_id' has been moved to behind the scenes.
* Moving the SqlCompatible behavior to the source and making it default behavior


Test/Case/MongodbTest.php
<?php






class MongodbTest extends PHPUnit_Framework_TestSuite {
    
/**
 * suite method, defines tests for this suite.
 *
 * @return void
 */
	public static function suite() {
		$suite = new PHPUnit_Framework_TestSuite('MongoDB Plugin tests');
        $casePath = CakePlugin::path('Mongodb') . DS . 'Test' . DS . 'Case';
		$suite->addTestFile($casePath . DS . 'Behavior' . DS . 'SqlCompatibleTest.php');
		$suite->addTestFile($casePath . DS . 'Behavior' . DS . 'SubCollectionValidatorBehaviorTest.php');
		$suite->addTestFile($casePath . DS . 'Datasource' . DS . 'MongodbSourceTest.php');
		$suite->addTestFile($casePath . DS . 'Model' . DS . 'MongodbAssociationsTest.php');
		$suite->addTestFile($casePath . DS . 'Model' . DS . 'MongodbInheritanceModelTest.php');
		return $suite;
	}
}




?>

Test/Case/MongoTestCase.php
<?php
/**
 * Base TestCase class for all mongoDB datasource tests.
 *
 * This datasource uses Pecl Mongo (http://php.net/mongo)
 * and is thus dependent on PHP 5.0 and greater.
 *
 * Copyright 2010, Yasushi Ichikawa http://github.com/ichikaway/
 *
 * Licensed under The MIT License
 * Redistributions of files must retain the above copyright notice.
 *
 * @copyright     Copyright (c) 2010, Yasushi Ichikawa http://github.com/ichikaway/
 * @package       mongodb
 * @subpackage    mongodb.tests.cases.model
 * @license       http://www.opensource.org/licenses/mit-license.php The MIT License
 */

/**
 * MongoDB Associactions test class
 *
 * @package       app
 * @subpackage    app.model
 * @property Model $Vehicle
 * @property Model $VehiclePart
 * @property Model $Car
 * @property Model $Engine
 */
class MongoTestCase extends CakeTestCase {
    
    public $debug = true;

/**
 * Database Instance
 *
 * @var MongodbSource
 * @access public
 */
	public $mongodb;
    
    /**
     *
     * @var boolean If true, test method data persists until next test
     */
    public $keepDataForNext = false;
    
    /**
     *
     * @var boolean If true, no data is dropped after each test (but is erased on setup)
     */
    public $keepAllData = false;

/**
 * Base Config
 *
 * @var array
 * @access public
 *
 */
	protected $_config = array(
		'datasource' => 'Mongodb.MongodbSource',
		'host' => 'localhost',
		'login' => '',
		'password' => '',
		'database' => 'test_mongo',
		'port' => 27017,
		'prefix' => '',
		//'persistent' => true,
	);
    
    public function __construct($name = NULL, array $data = array(), $dataName = '') {
        parent::__construct($name, $data, $dataName);
        if (!empty($_GET['debug'])) {
            $this->keepAllData = true;  
        }
    }
    
    
    
    ##### UTILITY METHODS #####
    
    
    
    function startTest($method){
        
    }
    function endTest($method) {
        if (!$this->keepDataForNext && !$this->keepAllData){
            $this->dropData();
        }
        $this->keepDataForNext = false;
    }

/**
 * Sets up the environment for each TEST METHOD
 *
 * @return void
 * @access public
 */
	public function setUp() {
        parent::setUp();
		//$config = ConnectionManager::enumConnectionObjects();
        $this->mongodb = ConnectionManager::getDataSource('test');
        $this->skipIf(!$this->mongodb instanceof MongodbSource, 'These tests must run on a MongoDB test database confirguration. Please set your "$test" database config to MongoDB.');
        $dbCon = new DATABASE_CONFIG();
        $this->_config = $dbCon->test;
	}

/**
 * Destroys the environment after each test method is run
 *
 * @return void
 * @access public
 */
	public function tearDown() {
	}


/**
 * get Mongod server version
 *
 * @return numeric
 * @access public
 */
	public function getMongodVersion() {
		$mongo = $this->Post->getDataSource();
		return $mongo->execute('db.version()');
	}

/**
 * Drop database
 *
 * @return void
 * @access public
 */
	public function dropData() {
        //$this->mongodb->connect();
        //debug($this->mongodb->MongoDB()); die();
        try {
			$db = $this->mongodb->getMongoDB(true);

			foreach($db->listCollections() as $collection) {
				$collection->drop();
			}
		} catch (MongoException $e) {
			trigger_error($e->getMessage());
		}
	}


    
    
    protected function assertArraySubsetMatches($result, $expected){
        if (!is_array($result) || !is_array($expected)) {
            $this->fail("assertArraySubsetMatches requires two arrays as arguments");
        }
        if (class_exists('Hash')) {
            $result = Hash::flatten($result);
            $expected = Hash::flatten($expected);
        } else {
            $result = Set::flatten($result);
            $expected = Set::flatten($expected);
        }
        foreach (array_keys($expected) as $key) {
            if (!key_exists($key, $result)) {
                if (!empty($_GET['debug'])){
                    debug($result); debug($expected); xdebug_print_function_stack(); ob_flush();
                }
                $this->fail("Key '{$key}' missing in resulting array. Enable debug output to inspect the arrays.");
            }
        }
        foreach (array_keys($result) as $key) {
            if (!key_exists($key, $expected)) {
                unset($result[$key]);
            }
        }
        $this->assertEquals($expected, $result);
    }
    
    public function debug($what){
        if (!empty($_GET['debug'])){
            debug($what);
            ob_flush();
        }
    }
    
    
    
    /**
     * Backwards compatible function for Hash::extract
     * 
     * @param array $set
     * @param string $path
     * @param string $compatPath
     * @return array Result of Hash::extract()
     */
    protected function _extract($set, $path, $compatPath = null){
        if (class_exists('Hash')) {
            return Hash::extract($set, $path);
        }
        else {
            return Set::extract($set, ($compatPath === null) ? $path : $compatPath);
        }
    }
}


Test/Case/Behavior/SqlCompatibleTest.php
<?php
/**
 * Tests specific to the sql compatible behavior
 *
 * PHP version 5
 *
 * Copyright (c) 2010, Andy Dawson
 *
 * Licensed under The MIT License
 * Redistributions of files must retain the above copyright notice.
 *
 * @filesource
 * @copyright     Copyright (c) 2010, Andy Dawson
 * @link          www.ad7six.com
 * @package       mongodb
 * @subpackage    mongodb.tests.cases.behaviors
 * @since         v 1.0 (14-Dec-2010)
 * @license       http://www.opensource.org/licenses/mit-license.php The MIT License
 */


App::uses('Model', 'Model');
App::uses('AppModel', 'Model');
App::uses('MongodbSource', 'Mongodb.Model/Datasource');
App::uses('MongodbAppModel', 'Mongodb.Model');


/**
 * SqlCompatiblePost class
 *
 * @uses          Post
 * @package       mongodb
 * @subpackage    mongodb.tests.cases.behaviors
 */
class SqlCompatiblePost extends MongodbAppModel {

/**
 * useDbConfig property
 *
 * @var string 'test_mongo'
 * @access public
 */
	public $useDbConfig = 'test_mongo';

/**
 * actsAs property
 *
 * @var array
 * @access public
 */
	public $actsAs = array(
		'Mongodb.SqlCompatible'
	);

	public $lastQuery = array();

	public function beforeFind($query) {
		$this->lastQuery = $query;
		return $query;
	}
}

/**
 * SqlCompatibleTest class
 *
 * @uses          CakeTestCase
 * @package       mongodb
 * @subpackage    mongodb.tests.cases.behaviors
 */
class SqlCompatibleTest extends CakeTestCase {

/**
 * Default db config. overriden by test db connection if present
 *
 * @var array
 * @access protected
 */
	protected $_config = array(
		'datasource' => 'Mongodb.MongodbSource',
		'host' => 'localhost',
		'login' => '',
		'password' => '',
		'database' => 'test_mongo',
		'port' => 27017,
		'prefix' => '',
		'persistent' => false,
	);

	public function setUp() {
		$connections = ConnectionManager::enumConnectionObjects();

		if (!empty($connections['test']['classname']) && $connections['test']['classname'] === 'mongodbSource') {
			$config = new DATABASE_CONFIG();
			$this->_config = $config->test;
		}

		if(!isset($connections['test_mongo'])) {
			ConnectionManager::create('test_mongo', $this->_config);
			$this->Mongo = new MongodbSource($this->_config);
		}

		$this->Post = ClassRegistry::init(array('class' => 'SqlCompatiblePost', 'alias' => 'Post', 'ds' => 'test_mongo'), true);
	}

/**
 * Sets up the environment for each test method
 *
 * @return void
 * @access public
 */
	public function startTest($method) {
		$this->_setupData();
	}

/**
 * Destroys the environment after each test method is run
 *
 * @return void
 * @access public
 */
	public function endTest($method) {
		$this->Post->deleteAll(true);
	}

	public function tearDown() {
		unset($this->Post);
		ClassRegistry::flush();
	}

/**
 * testNOT method
 *
 * @return void
 * @access public
 */
	public function testNOT() {
		$expected = array(1, 2, 3, 4, 5, 6, 7, 8, 9, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20);
		$result = $this->Post->find('all', array(
			'conditions' => array(
				'title !=' => 10,
			),
			'fields' => array('_id', 'title', 'number'),
			'order' => array('number' => 'ASC')
		));

		$result = $this->_extract($result, '{n}.Post.title');
		$this->assertEqual($expected, $result);

		$conditions = array(
			'title' => array('$ne' => 10)
		);
		$this->assertEqual($conditions, $this->Post->lastQuery['conditions']);

		$result = $this->Post->find('all', array(
			'conditions' => array(
				'NOT' => array(
					'title' => 10
				),
			),
			'fields' => array('_id', 'title', 'number'),
			'order' => array('number' => 'ASC')
		));
		$result = $this->_extract($result, '{n}.Post.title');
		$this->assertEqual($expected, $result);

		$conditions = array(
			'title' => array('$nin' => array(10))
		);
		$this->assertEqual($conditions, $this->Post->lastQuery['conditions']);
	}

/**
 * testGTLT method
 *
 * @return void
 * @access public
 */
	public function testGTLT() {
		$expected = array(8, 9, 10, 11, 12, 13);
		$result = $this->Post->find('all', array(
			'conditions' => array(
				'title >' => 7,
				'title <' => 14,
			),
			'fields' => array('_id', 'title', 'number'),
			'order' => array('number' => 'ASC')
		));
		$result = $this->_extract($result, '{n}.Post.title');
		$this->assertEqual($expected, $result);

		$conditions = array(
			'title' => array(
				'$gt' => 7,
				'$lt' => 14
			)
		);
		$this->assertEqual($conditions, $this->Post->lastQuery['conditions']);
	}

/**
 * testGTE method
 *
 * @return void
 * @access public
 */
	public function testGTE() {
		$expected = array(19, 20);
		$result = $this->Post->find('all', array(
			'conditions' => array(
				'title >=' => 19,
			),
			'fields' => array('_id', 'title', 'number'),
			'order' => array('number' => 'ASC')
		));
		$result = $this->_extract($result, '{n}.Post.title');
		$this->assertEqual($expected, $result);

		$conditions = array(
			'title' => array('$gte' => 19)
		);
		$this->assertEqual($conditions, $this->Post->lastQuery['conditions']);
	}

/**
 * testOR method
 *
 * @return void
 * @access public
 */
	public function testOR() {
		$expected = array(1, 2, 19, 20);
		$result = $this->Post->find('all', array(
			'conditions' => array(
				'OR' => array(
					'title <=' => 2,
					'title >=' => 19,
				)
			),
			'fields' => array('_id', 'title', 'number'),
			'order' => array('number' => 'ASC')
		));
		$result = $this->_extract($result, '{n}.Post.title');
		$this->assertEqual($expected, $result);

		$conditions = array(
			'$or' => array(
				array('title' => array('$lte' => 2)),
				array('title' => array('$gte' => 19))
			)
		);
		$this->assertEqual($conditions, $this->Post->lastQuery['conditions']);
	}


	/**
	 * Tests find method with conditions _id=>array()
	 *
	 * @return void
	 * @access public
	 */
	public function testFindConditionIn() {

		for ($i = 1; $i <= 5; $i++) {
			$data = array(
					'_id' => 'A1' . $i,
					'title' => $i,
					);
			$saveData['Post'] = $data;
			$this->Post->create();
			$this->Post->save($saveData);
		}

		$params = array('conditions' => array('_id' => array('A11', 'A12')));
		$result = $this->Post->find('all', $params);

		$expected = array('A11','A12');
		$result = $this->_extract($result, '{n}.Post._id');
		$this->assertEqual($expected, $result);
		$this->assertEqual(2, count($result));

		$conditions = array(
			'_id' => array(
				'$in' => array('A11', 'A12')
			)
		);
		$this->assertEqual($conditions, $this->Post->lastQuery['conditions']);


		$params = array('conditions' => array('_id' => array('$nin' => array('A11', 'A12'))));
		$result = $this->Post->find('all', $params);
		//$expected = array('A13','A14');
		$result = $this->_extract($result, '{n}.Post._id');
        //debug($result); exit();
		$this->assertTrue(in_array('A13', $result));
		$this->assertFalse(in_array('A11', $result));
		$this->assertFalse(in_array('A12', $result));
		$this->assertEqual(23, count($result));


		$conditions = array(
			'_id' => array(
				'$nin' => array('A11', 'A12')
			)
		);
		$this->assertEqual($conditions, $this->Post->lastQuery['conditions']);
	}


/**
 * Order method
 *
 * @return void
 * @access public
 */
	public function testOrderDESC() {
		$expected = array(20, 19);
		$result = $this->Post->find('all', array(
			'conditions' => array('title >' => 18),
			'fields' => array('_id', 'title'),
			'order' => array('title DESC')
		));
		$result = $this->_extract($result, '{n}.Post.title');
		$this->assertEqual($expected, $result);

		$order = array(array('title' => 'DESC'));
		$this->assertEqual($order, $this->Post->lastQuery['order']);
	}

/**
 * Order method
 *
 * @return void
 * @access public
 */
	public function testOrderASC() {
		$expected = array(19, 20);
		$result = $this->Post->find('all', array(
			'conditions' => array('title >' => 18),
			'fields' => array('_id', 'title'),
			'order' => array('title ASC')
		));
		$result = $this->_extract($result, '{n}.Post.title');
		$this->assertEqual($expected, $result);

		$order = array(array('title' => 'ASC'));
		$this->assertEqual($order, $this->Post->lastQuery['order']);
	}


/**
 * Order method with model alias
 *
 * @return void
 * @access public
 */
	public function testOrderWithModelAlias() {
		$expected = array(20, 19);
		$result = $this->Post->find('all', array(
			'conditions' => array('title >' => 18),
			'fields' => array('_id', 'title'),
			'order' => array('Post.title DESC')
		));
		$result = $this->_extract($result, '{n}.Post.title');
		$this->assertEqual($expected, $result);
	}



/**
 * setupData method
 *
 * @return void
 * @access protected
 */
	protected function _setupData() {
		$this->Post->deleteAll(true, false);
		$this->Post->primaryKey = '_id';

		for ($i = 1; $i <= 20; $i++) {
			$data = array(
				'title' => $i,
			);
			$saveData['Post'] = $data;
			$this->Post->create();
			$this->Post->save($saveData);
		}
	}
    
    
    /**
     * Backwards compatible function for Hash::extract
     * 
     * @param array $set
     * @param string $path
     * @param string $compatPath
     * @return array Result of Hash::extract()
     */
    protected function _extract($set, $path, $compatPath = null){
        if (class_exists('Hash')) {
            return Hash::extract($set, $path);
        }
        else {
            return Set::extract($set, ($compatPath === null) ? $path : $compatPath);
        }
    }
}


Model/Behavior/SchemalessBehavior.php
<?php
/**
 * Schemaless behavior.
 *
 * Adds functionality specific to MongoDB/schemaless dbs
 * Allow /not/ specifying the model's schema, and derive it (for cake-compatibility) from the data
 * being saved. Note that used carelessly this is a pretty dangerous thing to allow - means a user
 * can modify input forms adding whatever fields they like (unless you'er using the security
 * component) and fill your db with their junk.
 *
 * PHP version 5
 *
 * Copyright (c) 2010, Andy Dawson
 *
 * Licensed under The MIT License
 * Redistributions of files must retain the above copyright notice.
 *
 * @filesource
 * @copyright     Copyright (c) 2010, Andy Dawson
 * @link          www.ad7six.com
 * @package       mongodb
 * @subpackage    mongodb.models.behaviors
 * @since         v 1.0 (24-May-2010)
 * @license       http://www.opensource.org/licenses/mit-license.php The MIT License
 */

/**
 * SchemalessBehavior class
 *
 * @uses          ModelBehavior
 * @package       mongodb
 * @subpackage    mongodb.models.behaviors
 */
class SchemalessBehavior extends ModelBehavior {

/**
 * name property
 *
 * @var string 'Schemaless'
 * @access public
 */
	public $name = 'Schemaless';

/**
 * settings property
 *
 * @var array
 * @access public
 */
	public $settings = array();

/**
 * defaultSettings property
 *
 * @var array
 * @access protected
 */
	protected $_defaultSettings = array(
	);

/**
 * setup method
 *
 * Don't currently have any settings at all - disabled
 *
 * @param mixed $Model
 * @param array $config array()
 * @return void
 * @access public
 */

	public function setup(Model $model, $config = array()) {
		//$this->settings[$Model->alias] = array_merge($this->_defaultSettings, $config);
	}


/**
 * beforeSave method
 *
 * Set the schema to allow saving whatever has been passed
 *
 * @param mixed $Model
 * @return void
 * @access public
 */
	public function beforeSave(Model $Model) {
		$Model->cacheSources = false;
		$Model->schema(true);
		return true;
	}
}

Model/Behavior/SubCollectionValidatorBehavior.php
<?php
/**
 * SubCollectionValidatorBehavior.
 *
 * Enalbe validation of subsets, used in some NoSQL
 * Databases like MongoDB.
 *
 * PHP versions 5
 *
 * Copyright 2012, Radig Soluções em TI (http://radig.com.br)
 *
 * Licensed under The MIT License
 * Redistributions of files must retain the above copyright notice.
 *
 * @copyright     Copyright 2012, Radig Soluções em TI. (http://radig.com.br)
 * @link          http://github.com/radig/
 * @package       Mongodb.Model.Behavior
 * @license       MIT License (http://www.opensource.org/licenses/mit-license.php)
 */

App::uses('CakeValidationSet', 'Model/Validator');
class SubCollectionValidatorBehavior extends ModelBehavior {

    private $_methods = array();

    private $_Model = null;

    public function setup(Model $model, $config = array()) {
        if(!isset($this->_methods[$model->name])) {
            if (method_exists($model, 'validator')) {
                $this->_methods[$model->name] = $model->validator()->getMethods();
            }
            else {
                throw new MissingBehaviorException('SubCollectionValidator behvior not supported for CakePHP v'.Configure::version().'.');
            }
            
        }
    }

    public function beforeValidate(Model $model) {
        $this->_Model = $model;

        if(is_a($model->getDataSource(), 'Mongodbsource') && isset($model->collectionValidate)) {
            foreach($model->data[$model->alias] as $fieldName => $value) {
                if(!isset($model->collectionValidate[$fieldName])) {
                    continue;
                }

                $this->_validateCollection($model->data[$model->alias][$fieldName], $model->collectionValidate[$fieldName]);
            }
        }

        return true;
    }

    public function beforeSave(Model $model) {
        return empty($model->validationErrors);
    }

    protected function _validateCollection($data, $ruleset) {
        foreach($data as $field => $value) {
            if(is_array($value) && isset($ruleset["_{$field}"])) {
                $status &= $this->_validateCollection($data[$field], $ruleset["_{$field}"]);
                continue;
            }

            if(!isset($ruleset[$field])) {
                continue;
            }

            $obj = new CakeValidationSet($field, $ruleset[$field]);
            $obj->setMethods($this->_methods[$this->_Model->name]);
            $errors = $obj->validate($data);

            foreach($errors as $error) {
                $this->_Model->invalidate($field, $error);
            }
        }
    }
}

Model/MongodbAppModel.php
<?php
/**
 * Mongodb App Model
 *
 *
 * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
 * Copyright 2005-2010, Cake Software Foundation, Inc. (http://cakefoundation.org)
 *
 * Licensed under The MIT License
 * Redistributions of files must retain the above copyright notice.
 *
 * @copyright     Copyright 2013, DizyArt http://github.com/dizyart/
 * @link          http://cakephp.org
 * @package       Mongodb
 * @subpackage    Mongodb.Model
 * @since         Mongodb 2.0a
 * @license       MIT License (http://www.opensource.org/licenses/mit-license.php)
 * @todo Possibly move the Inheritance methods to a separate class?
 **/

App::uses('AppModel', 'Model');

/**
 * Application model for Mongodb.
 *
 * Adds some necessary adaptations for saving and retreiving data.
 *
 * @package       Mongodb.Model
 */
class MongodbAppModel extends AppModel {
    
    protected $_modelParentsCache = null;
    protected $_classParentsCache = null;
    
    public $parentModels = array();
    /**
     *
     * @var string ClassName of the parent model which holds the inherited table name ($useTable which is inherited)
     */
    public $tableParent = null;
    
    protected $_baseModelClass = array('AppModel', 'MongodbAppModel', 'Model');
    
    
    public function __construct($id = false, $table = null, $ds = null) {
        $this->_collectParentProperties();
        $this->useTable = $this->resolveUseTable();
        parent::__construct($id, $table, $ds);
    }
    
    
    private function _collectParentProperties() {
        foreach ($this->getModelParents() as $parentModel) {
            $this->parentModels[$parentModel] = get_class_vars($parentModel);
        }
    }
    
    public function resolveUseTable() {
        if ($this->useTable !== null) {
            return $this->useTable;
        }
        $useTable = null;
        $tableParent = null;
        foreach($this->parentModels as $class => $vars){
            if (isset($vars['useTable'])){
                $useTable = $vars['useTable'];
                $tableParent = $class;
            }
        }
        if ($useTable === null && count($this->parentModels)>0) {
            $firstParent = reset($this->parentModels);
            $tableParent = key($this->parentModels);
            $useTable = (!empty($firstParent['name'])) ? Inflector::tableize($firstParent['name']) : Inflector::tableize($tableParent);
        }
        $this->tableParent = $tableParent;
        return $useTable;
    }
    
    public function getTableParent(){
        if (!empty($this->tableParent)){
            return $this->tableParent;
        }
        else {
            return get_class($this);
        }
    }

/**
 * Adapter for Model::_saveMulti
 * 
 * Method is overriden because saving HABTM data structure cannot be properly recognized by
 * Model class. Specifically, the MongoId (or a string representation) is not recognized 
 * by cake as a valid ID type, hence skipping it in the ID check:
 * 
 *      if ((is_string($row) && (strlen($row) == 36 || strlen($row) == 16)) || is_numeric($row))
 *
 * @param array $joined Data to save
 * @param integer|string $id ID of record in this model
 * @param DataSource $db
 * @return void
 * @todo Reduce the code to minimum to ensure compatibility with other Cake versions
 * @todo write tests for other data structures, increase code coverage
 */
	protected function _saveMulti($joined, $id, $db) {
		foreach ($joined as $assoc => $data) {

			if (isset($this->hasAndBelongsToMany[$assoc])) {
				list($join) = $this->joinModel($this->hasAndBelongsToMany[$assoc]['with']);

				if ($with = $this->hasAndBelongsToMany[$assoc]['with']) {
					$withModel = is_array($with) ? key($with) : $with;
					list(, $withModel) = pluginSplit($withModel);
					$dbMulti = $this->{$withModel}->getDataSource();
				} else {
					$dbMulti = $db;
				}
                if (method_exists($this->{$join}, '_isUUIDField')){
                    $isUUID = !empty($this->{$join}->primaryKey) && $this->{$join}->_isUUIDField($this->{$join}->primaryKey);
                }
                else {
                    $isUUID = false;
                }

				$newData = $newValues = $newJoins = array();
				$primaryAdded = false;

				$fields = array(
					$dbMulti->name($this->hasAndBelongsToMany[$assoc]['foreignKey']),
					$dbMulti->name($this->hasAndBelongsToMany[$assoc]['associationForeignKey'])
				);

				$idField = $db->name($this->{$join}->primaryKey);
				if ($isUUID && !in_array($idField, $fields)) {
					$fields[] = $idField;
					$primaryAdded = true;
				}

				foreach ((array)$data as $row) {
					if ((is_string($row) && (strlen($row) == 24)) || (is_object($row) && $row instanceof MongoId) ) {
						$newJoins[] = $row;
						$values = array($id, $row);
						if ($isUUID && $primaryAdded) {
							$values[] = String::uuid();
						}
						$newValues[$row] = $values;
						unset($values);
					} elseif (isset($row[$this->hasAndBelongsToMany[$assoc]['associationForeignKey']])) {
						if (!empty($row[$this->{$join}->primaryKey])) {
							$newJoins[] = $row[$this->hasAndBelongsToMany[$assoc]['associationForeignKey']];
						}
						$newData[] = $row;
					} elseif (isset($row[$join]) && isset($row[$join][$this->hasAndBelongsToMany[$assoc]['associationForeignKey']])) {
						if (!empty($row[$join][$this->{$join}->primaryKey])) {
							$newJoins[] = $row[$join][$this->hasAndBelongsToMany[$assoc]['associationForeignKey']];
						}
						$newData[] = $row[$join];
					}
				}

				$keepExisting = $this->hasAndBelongsToMany[$assoc]['unique'] === 'keepExisting';
				if ($this->hasAndBelongsToMany[$assoc]['unique']) {
					$conditions = array(
						$join . '.' . $this->hasAndBelongsToMany[$assoc]['foreignKey'] => $id
					);
					if (!empty($this->hasAndBelongsToMany[$assoc]['conditions'])) {
						$conditions = array_merge($conditions, (array)$this->hasAndBelongsToMany[$assoc]['conditions']);
					}
					$associationForeignKey = $this->{$join}->alias . '.' . $this->hasAndBelongsToMany[$assoc]['associationForeignKey'];
					$links = $this->{$join}->find('all', array(
						'conditions' => $conditions,
						'recursive' => empty($this->hasAndBelongsToMany[$assoc]['conditions']) ? -1 : 0,
						'fields' => $associationForeignKey,
					));

					$oldLinks = (class_exists('Hash')) 
                        ? Hash::extract($links, "{n}.{$associationForeignKey}")
                        : Set::extract($links, "{n}.{$associationForeignKey}");
					if (!empty($oldLinks)) {
						if ($keepExisting && !empty($newJoins)) {
							$conditions[$associationForeignKey] = array_diff($oldLinks, $newJoins);
						} else {
							$conditions[$associationForeignKey] = $oldLinks;
						}
						$dbMulti->delete($this->{$join}, $conditions);
					}
				}

				if (!empty($newData)) {
					foreach ($newData as $data) {
						$data[$this->hasAndBelongsToMany[$assoc]['foreignKey']] = $id;
						if (empty($data[$this->{$join}->primaryKey])) {
							$this->{$join}->create();
						}
						$this->{$join}->save($data);
					}
				}

				if (!empty($newValues)) {
					if ($keepExisting && !empty($links)) {
						foreach ($links as $link) {
							$oldJoin = $link[$join][$this->hasAndBelongsToMany[$assoc]['associationForeignKey']];
							if (!in_array($oldJoin, $newJoins)) {
								$conditions[$associationForeignKey] = $oldJoin;
								$db->delete($this->{$join}, $conditions);
							} else {
								unset($newValues[$oldJoin]);
							}
						}
						$newValues = array_values($newValues);
					}
					if (!empty($newValues)) {
						$dbMulti->insertMulti($this->{$join}, $fields, $newValues);
					}
				}
			}
		}
	}
    
    public function getParentModelInstance(){
        $className = $this::getParentModelName();
        return ClassRegistry::init(array('class' => $className), true);
    }
    
    public function hasParentModel(){
        $parents = $this->getModelParents();
        return !empty($parents);
    }
    /**
     * Get the parent models this model inherits. 
     * 
     * If you plan on using an intermediate class between AppModel and your Model classes, 
     * add the name of the base model class to $this->_
     * 
     * @return array Parent classes of Model type
     */
    public function getModelParents(){
        if ($this->_modelParentCache !== null) return $this->_modelParentCache;
        $parents = $this->getClassParents();
        $return = array();
        if (is_string($this->_baseModelClass)) {
                $this->_baseModelClass = array($this->_baseModelClass);
            }
        $stop_at = array_merge((array)$this->_baseModelClass, array('AppModel', 'Model', 'MongoAppModel'));
        foreach($parents as $parent) {
            
            if (in_array($parent, $stop_at)){
                break;
            }
            array_push($return, $parent);
        }
        $this->_modelParentCache = $return;
        return $return;
    }
    
    public static function getParentModelName(){
        $parents = $this->getModelParents();
        
    }
    
    public final function getClassParents($class=null){
        if ($this->_classParentsCache === null) {
            $this->_classParentsCache = array();
            $_class = (is_object($class) || is_string($class) && class_exists($class))
                ? $class
                : get_class($this);

            if (function_exists('class_parents')) {
                $this->_classParentsCache = array_values(class_parents($_class));
            }
            else {
                while($parent = get_parent_class($_class)){
                    array_push($this->_classParentsCache, $parent);
                    $_class = $parent;
                }
            }
        }
        return $this->_classParentsCache;
    }
    
}
